rfd/0197-prometheus-metrics.md

```diff
@@ -0,0 +1,226 @@
+
+#### Do
+
+- Take the local in-process registry as an argument in your service constructor/main routine, like you would receive a
+  logger, and register your metrics against it.
+- Pass the registry as a `promtetheus.Registerer`
+
+```golang
+func NewService(log *slog.Logger, reg prometheus.Registerer) (Service, error) {
+    myMetric := prometheus.NewGauge(prometheus.GaugeOpts{
+        Namespace: teleport.MetricNamespace,
+		Subsystem: metricsSubsystem,
+        Name: "my_metric",
+        Help: "Measures the number of foos doing bars.",
+    }),
+    if err := reg.Register(myMetric); err != nil {
+		return nil, trace.Wrap(err, "registering metric")
+    }
+    // ...
+}
+```
+
+#### Don't
+
+- Register against the global prometheus registry
+  ```golang
+  func NewService(log *slog.Logger) (Service, error) {
+      myMetric := prometheus.NewGauge(prometheus.GaugeOpts{
+          Namespace: teleport.MetricNamespace,
+  		Subsystem: metricsSubsystem,
+          Name: "my_metric",
+          Help: "Measures the number of foos doing bars.",
+      }),
+      if err := prometheus.Register(myMetric); err != nil {
+  		return nil, trace.Wrap(err, "registering metric")
+      }
+      // ...
+  }
+  ```
+
+- Pass the registry as a `*prometheus.Registry`
+
+  ```golang
+  func NewService(log *slog.Logger, reg *prometheus.Registry) (Service, error) {
+      myMetric := prometheus.NewGauge(prometheus.GaugeOpts{
+          Namespace: teleport.MetricNamespace,
+  		Subsystem: metricsSubsystem,
+          Name: "my_metric",
+          Help: "Measures the number of foos doing bars.",
+      }),
+      if err := reg.Register(myMetric); err != nil {
+  		return nil, trace.Wrap(err, "registering metric")
+      }
+      // ...
+  }
+  ```
+
+### Storing metrics
+
+#### Do
+
+- Store metrics in a private struct in your package
+
+```golang
+type metrics struct {
+    currentFoo prometheus.Gauge
+    barCounter *prometheus.CounterVec
+}
+
+func newMetrics(reg prometheus.Registerer) (*metrics, error) {
+    m := metrics{
+        currentFoo: prometheus.NewGauge(prometheus.GaugeOpts{
+            Namespace: teleport.MetricNamespace,
+	    	Subsystem: metricsSubsystem,
+            Name: "foo_current",
+            Help: "Measures the number of foos.",
+        }),
+        barCounter: // ...
+    } 
+    
+    errs := trace.NewAggregate(
+        reg.Register(m.currentFoo),
+        reg.Register(m.barCounter),
+    )
+    
+    if errs != nil {
+        return trace.Wrap(err, "registering metrics")
+    }
+    
+    return &m
+}
+
+```
+
+#### Don't
+
+- Store metrics in a package-scoped variable
+
+```golang
+var (
+    currentFoo = prometheus.NewGauge(prometheus.GaugeOpts{
+            Namespace: teleport.MetricNamespace,
+	    	Subsystem: metricsSubsystem,
+            Name: "foo_current",
+            Help: "Measures the number of foos.",
+        })
+    // ...
+)
+```
+
+### Naming metrics
+
+#### Do
+
+- Use `teleport.MetricsNamespace` as the namespace.
```

Comment:
Do we want this to be true for every executable? Should tbot metrics be prefixed by `tbot`?

Reply:
I can see this being nice from a product PoV - as a customer I can always rely on Teleport-related stuff landing in the "teleport" namespace.

The counterpoint is that other binaries get worse subsystems - we'd have to do "tbot-sub1", "tbot-sub2", etc. It's not really a problem if "tbot" never sub-divides its metrics, for example.

Maybe, as a rule of thumb, use the default namespace but the rule could be lifted case-by-case (for example for a significantly "large" 3rd binary).